File: python/circular_queue.py
class CircularQueue:
    def __init__(self, k):
        self.k = k
        self.queue = [None] * k
        self.head = self.tail = -1

    def enqueue(self, value):
        if ((self.tail + 1) % self.k == self.head):
            return False
        if (self.head == -1):
            self.head = 0
        self.tail = (self.tail + 1) % self.k
        self.queue[self.tail] = value
        return True

    def dequeue(self):
        if (self.head == -1):
            return False
        if (self.head == self.tail):
            self.head = self.tail = -1
        else:
            self.head = (self.head + 1) % self.k
        return True

if __name__ == "__main__":
    cq = CircularQueue(5)
    cq.enqueue(10)
    cq.enqueue(20)
    cq.enqueue(30)
    print("Enqueued items. Head is at index:", cq.head)

File: java/CircularQueue.java
public class CircularQueue {
    int[] queue;
    int front, rear, size, capacity;

    public CircularQueue(int capacity) {
        this.capacity = capacity;
        queue = new int[capacity];
        front = -1;
        rear = -1;
        size = 0;
    }

    public void enqueue(int item) {
        if (size == capacity) return;
        if (front == -1) front = 0;
        rear = (rear + 1) % capacity;
        queue[rear] = item;
        size++;
    }

    public int dequeue() {
        if (size == 0) return -1;
        int item = queue[front];
        front = (front + 1) % capacity;
        size--;
        if (size == 0) { front = -1; rear = -1; }
        return item;
    }

    public static void main(String[] args) {
        CircularQueue cq = new CircularQueue(5);
        cq.enqueue(10);
        cq.enqueue(20);
        cq.enqueue(30);
        System.out.println(cq.dequeue() + " dequeued");
    }
}
